### PR TITLE
Add k-means++ initialization.

### PR DESCRIPTION
**Pull request by rcurtin**
I implemented the k-means++ initialization strategy for k-means like three years ago but for some reason never got around to contributing it back upstream.  So, here it is.

https://en.wikipedia.org/wiki/K-means%2B%2B

It's quite an effective initialization strategy, and is often used in practice.  It seems to often outperform the Bradley/Fayyad refined start strategy that we have implemented.

### CODE REVIEW

File: src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
+/**
+ * @file kmeans_plus_plus_initialization.hpp
```
Comment (zoq) [inline src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp:2]:
```suggestion
 * @file methods/kmeans/kmeans_plus_plus_initialization.hpp
```

File: src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+              mlpack::metric::SquaredEuclideanDistance::Evaluate(data.col(p),
+              centroids.col(j));
+          minDistance = std::min(distance, minDistance);
+        }
+
+        distribution[p] = minDistance;
+      }
+
+      // Next normalize the distribution (actually technically we could avoid
+      // this.)
```
Comment (zoq) [inline src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp:85]:
```suggestion
      // this).
```

File: src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+      // Turn it into a CDF for convenience...
+      for (size_t j = 1; j < distribution.n_elem; ++j)
+        distribution[j] += distribution[j - 1];
+
+      // Sample a point...
+      const double sampleValue = mlpack::math::Random();
+      double* elem = std::lower_bound(distribution.begin(), distribution.end(),
+          sampleValue);
+      size_t position = (size_t) (elem - distribution.begin()) / sizeof(double);
```
Comment (zoq) [inline src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp:96]: Perhaps we should use `const` here as well, to help the compiler.

File: src/mlpack/tests/main.cpp [hunk 4]
```diff
@@ -22,8 +22,8 @@ int main(int argc, char** argv)
    * each run.  This is good for ensuring that a test's tolerance is sufficient
    * across many different runs.
    */
-  // size_t seed = std::time(NULL);
-  // mlpack::math::RandomSeed(seed);
+  size_t seed = std::time(NULL);
+  mlpack::math::RandomSeed(seed);
```
Comment (zoq) [inline src/mlpack/tests/main.cpp:26]:
```suggestion
  // size_t seed = std::time(NULL);
  // mlpack::math::RandomSeed(seed);
```

File: src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp (hunk 2 above)
Comment (rcurtin) [reply src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp]: Nice catch, thanks!

File: src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp (hunk 1 above)
Comment (rcurtin) [reply src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp]: Ahh, nice, thank you.  I need to update the header guard too.

File: src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp (hunk 3 above)
Comment (rcurtin) [reply src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp]: Good point; done in 79d2621a.

File: src/mlpack/tests/main.cpp (hunk 4 above)
Comment (rcurtin) [reply src/mlpack/tests/main.cpp]: Thanks!  Forgot to recomment that after testing.

**Comment by rcurtin** [comment]
Oops, I forgot to run `[KMeansMainTest]` locally, but the issue should be resolved now. :+1:

**Review by zoq — COMMENTED** [review]
Looks like it would make sense to slightly increase the threshold for the kmeans test.

**Comment by rcurtin** [comment]
You're right---I ran the test 1000 times locally and it failed a decent number of times.  I reset the tolerance to be something a decent bit larger than the largest value I saw.  So, hopefully, we should never see this test fail in practice. :)

**Review by zoq — APPROVED** [review]
Looks great to me.

File: src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      // This computation is ripe for speedup with trees!  I am not sure exactly
+      // how much we would need to approximate, but I think it could be done
+      // without breaking the O(log k)-competitive guarantee (I think).
+      for (size_t p = 0; p < data.n_cols; ++p)
+      {
+        double minDistance = std::numeric_limits<double>::max();
+        for (size_t j = 0; j < i; ++j)
+        {
+          const double distance =
+              mlpack::metric::SquaredEuclideanDistance::Evaluate(data.col(p),
```
Comment (shrit) [inline src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp:76]: I am not sure why, but it is showing as there is an issue with indentation, probably it is my browser.

**Review by shrit — APPROVED** [review]

File: src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp (hunk 5 above)
Comment (shrit) [reply src/mlpack/methods/kmeans/kmeans_plus_plus_initialization.hpp:76]: It is my browser

**Pull request merged**